### DISCUSSION TITLE
Raise E2 Event

### DISCUSSION THREAD

**Discussion by LVancina** [srsran/srsRAN_Project#559] [General Help]
Greetings,

I am doing some experimentation with xApps, using the flexRIC framework, and I am trying to figure out how I can notify my xApp every time an initial UE message is received by srsRAN (i.e. handle_initial_ue_message() is invoked). Could anyone point me at an example that might be helpful? Thanks!

**Comment by pgawlowicz** [comment]
Hi, currently it is not supported by our current E2 agent implementation.  Also, I do not think it is supported on the FlexRIC side.

**Reply by pgawlowicz** [reply]
In the ORAN TS `O-RAN.WG3.RICARCH-R003-v05.00.docx` in Section 6.2.1, you can find info about

1.  SDL (Shared Data Layer): This functionality enables the Near-RT RIC platform to provide SDL services for xApps, which can be used to subscribe to database notifications and to read, write and modify information stored on the database. UE-NIB, R-NIB and other use case specific information may be exposed using the SDL services

2.  UE-NIB (Network Information Base): This functionality enables the Near-RT RIC platform to maintain UE related information:
-	Maintaining a list of UEs and associated data;
-	Tracking and correlation of the UE identities associated with the connected E2 Nodes.
Such information may be generated and accessed by the Near-RT RIC platform or authorized xApps


But it is not clear how those databases are populated with the UE information. 

A xApp can use `SDL API Procedures` (section 9.5) to request and subscribe for data from SDL. But I am not sure whether it is implemented by any open-source RIC platform.

**Reply by LVancina** [reply]
That is very interesting. It sounds like you are saying that a SDL has not been implemented for srsRAN yet and it is unclear from the ORAN technical specification how one might go about doing that. Am I understanding correctly?

Also, just from studying the code (which I don't fully understand yet), I am inferring that, in order to make certain data accessible to an xApp, I will first have to create an E2SM module and add it as an E2SM service. Is that an accurate assessment? It's starting to look like a laborious process to implement an xApp that requires more data than basic KMPs. Any recommendations?

**Reply by pgawlowicz** [reply]
I believe that SDL is part of RIC platform, so it does not need to be implemented on the srsRAN side. 
It is unclear, how SDL and UE-NIB should be filled.  In `O-RAN.WG3.RICARCH-R003-v05.00.docx`, you can find two use-cases related to getting information about the connected E2 nodes, but nothing about the connected UEs. Specifically:
- 9.5.8.1	Subscribe/Notify for E2NodeInfo and E2NodeList
- 9.5.8.2	Subscribe/Push for E2NodeInfo and E2NodeList

and those two structures are:
-	E2NodeList: List of known E2 Node ID and the associated E2 Node State
-	E2NodeInfo: E2 Node component configuration list and RAN function list

Regarding the second question, I am not sure if you are talking about the implementation of E2 agent or xApps. 
If you are interested, you can check our simplified ORAN-SC-RIC version, where we provide two KPM xApps, that work perfectly with our E2 Agent. 
You can access it here:
https://github.com/srsran/oran-sc-ric/

**Reply by LVancina** [reply]
@pgawlowicz Yes, I forgot that the SDL is a RIC component, not part of the RAN. I have not been able to verify that FlexRIC has implemented a SDL, which may be a major problem for my project.

In the mean time, I am just trying to figure out how information and control messages are sent to / received from the E2 agent(s) in srsRAN as well as how I can start implementing a custom E2 service model. If you or anyone else can point me in the right direction on that, then I would be very grateful.

**Reply by pgawlowicz** [reply]
You can find the E2 code here:
- https://github.com/srsran/srsRAN_Project/tree/main/include/srsran/e2
- https://github.com/srsran/srsRAN_Project/tree/main/lib/e2

Probably, you can have a closer look at KPM and RC modules:
- https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_impl.cpp
- https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_rc/e2sm_rc_impl.cpp

**Reply by pgawlowicz** [reply]
I would recommend putting intended your functionality into one of the existing modules (KPM or RC).
